### ISSUE TITLE
Defining design guidelines based on CSS support

### ISSUE THREAD

**Issue by JayPanoz** [readium/readium-css#139]
I'm submitting a topic for discussion, which goal is to agree on design guidelines that can help make choices and take resolutions.

**Short description of the issue/suggestion:**

While we did our best to [address EPUB Compatibility](https://readium.org/readium-css/docs/CSS21-epub_compat.html) through documentation, we didn’t necessarily address what ReadiumCSS should do in regards to web/engine/platform compatibility. 

It was acceptable to handle things on their own a few years ago e.g. hyphenation was not available in Chromium/Blink on Linux nor Windows, but that didn’t prevent us from using it in ReadiumCSS, especially as it’s kinda critical for published works. With new specs and updates of existing ones, it’s become a little bit more difficult though. My hope is that design guidelines or rules may help keep ReadiumCSS more consistent. 

As usual with CSS, things are not as clear-cut as with other languages, mainly because if the rendering engine doesn’t support something, it doesn’t break or err. I’ll try my best to make it as clear as possible.

1. what should we do with non-standard properties i.e. vendor-prefixed properties that are not on the standardisation track?
2. is progressive enhancement desirable, or should we keep it the most consistent possible by targeting the common denominator/baseline?

## Non-standard props

There isn’t a lot of such properties in ReadiumCSS, fortunately, but they may end in self-inflicted pain. Such is the case in #113, with a webkit property in safeguards, to protect against raised capitals (initial letter that is bigger but not dropped) w/o a specified `line-height`. That is consistent with iBooks.app, but creates an inconsistency in ReadiumCSS-based apps across platforms. In hindsight maybe it would be preferable to remove it, and keep things consistent in the Readium ecosystem. 

Worse still, in terms of consistency, it is difficult to explain why this is present in ReadiumCSS, but `-webkit-column-axis` and
`-webkit-column-progression` aren’t, while they could help with issues such as #96.

**Proposed Resolution:** I guess this is something we can create a design rule for e.g. “Don’t use props that are not on track to become standard.”

## Progressive enhancement

This is probably going to be more of a range – What is preferable? – than picking between 2 options – Y/N – since we wouldn’t have hyphenation in the first place.

I’ve started sharing an overview of new and updated specs in #137 and it’s kinda obvious there may be useful features we could use, but they are obviously supported from specific versions of each rendering engine, which comes with its own set of issues:

- it creates extra rendering inconsistencies on the same platform, using different versions;
- it may create noise for implementers in the form of additional bug reports;
- it may solve issues but only for the latest versions;
- probably more I can’t tell off the top of my head.

On the other hand, that could also help future-proof ReadiumCSS, which is not an insignificant positive.

**Proposed Resolution:** that probably can’t be a rule, but rather a process i.e. documenting how it could help, what the impacts would be, discussing/voting, and explaining why it was used or not. However, there is probably a way to narrow things down a little bit, before entering this process phase – which sounds extensive and wouldn’t make things really efficient if we have to discuss everything. 

Maybe restricting progressive enhancements to minor bugs, or targeting a max version of engines in addition to the min version (see #131). I’m open to ideas and others’ point of view here.

**Comment by mickael-menu** [comment]
I agree that there's value in a consistent base Readium CSS distribution. Maybe the progressive enhancement could be some kind of opt-in as separate stylesheets, similar to the [EBPAJ fonts patch](https://github.com/readium/readium-css/blob/master/css/dist/ReadiumCSS-ebpaj_fonts_patch.css)?

>As usual with CSS, things are not as clear-cut as with other languages, mainly because if the rendering engine doesn’t support something, it doesn’t break or err.

I know that you have a test suite in the repo. Is there a way to run them automatically? That could be super helpful to have automated tests with screenshot diffs, running on a device farm like BrowserStack to test many different browser in one go.

https://medium.com/seminal/useful-css-regression-testing-tools-99f09a7b8e4c
